src/canvas/graphic.ts

```diff
@@ -20,9 +20,9 @@ import { REDARAW_BIT, SHAPE_CHANGED_BIT } from '../graphic/constants';
 const pathProxyForDraw = new PathProxy(true);
 
 // Not use el#hasStroke because style may be different.
-function styleHasStroke(style: PathStyleProps) {
+function styleHasStroke(style: PathStyleProps, excludeLineWidth?: boolean) {
```

Comment:
I think it will have better performance to use another function that not consider `lineWidth`. See "Optimizing Sorting - Argument Adaptation" section in the [Maybe you don't need Rust and WASM to speed up your JS](https://mrale.ph/blog/2018/02/03/maybe-you-dont-need-rust-to-speed-up-your-js.html) article.

Reply:
I'm not sure. The reason why I want to exclude line width is that some legend-related cases can't pass. Please run VRTT to see the broken cases.

Reply:
And thanks for sharing the awesome article. It's very helpful.

Reply:
> I'm not sure. The reason why I want to exclude line width is that some legend-related cases can't pass. Please run VRTT to see the broken cases.

This is a bit of a surprise. Do you mind giving one fail test name so I don't have to switch from my current branch(which has a lot of breaks) and run all test cases

Reply:
`area-stack`, `area2`, `axis`, `axis-interval2`, `axisPosition` etc.

Reply:
Figured out why. It happens when switching from a style `{ stroke: '#ccc', lineWidth: 0}` to a another `{ stroke: '#ccc', lineWidth: 2}`.  Previous style didn't set `strokeStyle` because `lineWidth` is 0. Then the following style also didn't set it because it compares the `stroke` property with the previous and thought the `strokeStyle` doesn't change. 

To simplify the logic, we should ignore the factor of `lineWidth`.

Reply:
I'm glad we found the reason. I just found another potential warning `Expected color but found '['` when color is gradient. So I check if stroke/fill color is valid before setting strokeStyle/fillStyle.

Reply:
Saw this explanation right after I commented about why doing the string check😃